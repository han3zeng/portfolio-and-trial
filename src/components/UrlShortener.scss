.container {
  max-width: 600px;
  margin: 30px auto 30px auto;
}

.inputPanel {
  margin: 30px 0;
}

.input {
  border: none;
  border-bottom: 1px solid #333;
  display: block;
  width: 100%;
  outline: none;
  margin: 10px 0 30px 0;
  padding: 5px;
  font-size: 16px;
}

.button {
  display: block;
  border: none;
  padding: 0;
  margin: 0;
  width: 100%;
  max-width: 130px;
  padding: 10px 12px;
  background-color: #333;
  color: #FFD700;
  font-size: 20px;
  cursor: pointer;
  outline: none;
}

.response {
  margin: 30px 0;
}

.error {
  color: #D8000C;
  font-size: 14px;
}

.result {
  font-size: 18px;
  .shortenUrl {
    margin-top: 10px;
    background-color: #f2f2f2;
    padding: 6px 8px;
    margin-left: 10px;
    display : inline-block;
  }
}
