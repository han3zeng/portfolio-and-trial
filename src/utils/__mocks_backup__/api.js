
/*
  this is just the file to demonstrate jest-manual-mock
  now, the mock-createShortUrl is implemented in test file
*/
const createShortUrl = async (bodyObj) => {
  const res = await Promise.resolve({
    data: {
      shortenUrl: 'test-shorten-url',
    },
  });
  return res;
};

export default {
  createShortUrl,
};

export {
  createShortUrl,
};
