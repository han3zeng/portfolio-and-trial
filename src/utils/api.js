import config from '../../config/config';

const { apiOrigin } = config;

const postOption = (body) => Object.freeze({
  method: 'POST',
  mode: 'cors',
  cache: 'no-cache',
  credentials: 'omit',
  headers: {
    'Content-Type': 'application/json',
  },
  redirect: 'follow',
  referrer: 'no-referrer',
  body: JSON.stringify(body),
});


const asyncResolver = async (promise) => {
  const res = {
    data: null,
    error: null,
  };
  try {
    const response = await promise;
    res.data = await response.json();
    return res;
  } catch (e) {
    res.error = e;
    return res;
  }
};

const createShortUrl = async (bodyObj) => {
  const res = await asyncResolver(fetch(`${apiOrigin}/api/shorturl/new`, postOption(bodyObj)));
  return res;
};

const fetchUserProfile = async (accessToken) => {
  const options = {
    method: 'POST',
    mode: 'cors',
    cache: 'no-cache',
    credentials: 'include',
    headers: {
      'Content-Type': 'application/json',
      'Authorization': `Bearer ${accessToken}`,
    },
    redirect: 'follow',
    referrer: 'no-referrer',
  };
  const res = await asyncResolver(fetch(`${apiOrigin}/api/retrieve-user-profile`, options));
  return res;
};

export default {
  createShortUrl,
  fetchUserProfile,
};

export {
  createShortUrl,
  fetchUserProfile,
};
