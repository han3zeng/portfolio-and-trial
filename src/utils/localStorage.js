const USER_PROFILE = 'userProfile';
const AUTHORIZED = 'authorized';

const keys = [
  USER_PROFILE,
  AUTHORIZED
];

const response = ({ ok, status, data }) => {
  return Object.freeze({
    ok,
    status,
    data,
  });
};


// TODO: window check - decoration
function AuthContextManager() {
  const fetchAuthContext = () => {
    const userProfile = localStorage.getItem(USER_PROFILE);
    const authorized = localStorage.getItem(AUTHORIZED);
    if (window && userProfile && authorized) {
      return response({
        ok: true,
        status: 1,
        data: {
          userProfile: JSON.parse(userProfile),
          authorized,
        },
      });
    }
    return response({
      ok: false,
      status: 0,
      data: null,
    });
  };


  const setAuthContext = ({ userProfile, authorized }) => {
    if (!userProfile || !authorized) {
      return response({
        ok: false,
        status: 0,
        data: null,
      });
    }
    localStorage.setItem(USER_PROFILE, JSON.stringify(userProfile));
    localStorage.setItem(AUTHORIZED, authorized);
    return response({
      ok: true,
      status: 1,
      data: null,
    });
  };


  const clearAtuhContext = () => {
    localStorage.removeItem(USER_PROFILE);
    localStorage.removeItem(AUTHORIZED);
    return response({
      ok: true,
      status: 1,
      data: null,
    });
  };

  return {
    fetchAuthContext,
    setAuthContext,
    clearAtuhContext,
  };
}

export {
  AuthContextManager,
};
