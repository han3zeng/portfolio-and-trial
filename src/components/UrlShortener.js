import React, { useState } from 'react';
import styles from './UrlShortener.scss';
import { createShortUrl } from '../utils/api';
import withStyles from 'isomorphic-style-loader/withStyles';


const UrlShortener = () => {
  const [errorState, setError] = useState(false);
  const [shortenUrl, setShortenUrl] = useState(undefined);
  const [url, setUrl] = useState('');
  const onInputChange = (e) => {
    setUrl(e.target.value);
  };
  const submitHandler = async () => {
    const originalUrl = url;
    const result = await createShortUrl({ originalUrl });
    if (result.error) {
      setError(true);
    } else {
      setShortenUrl(result.data.shortenUrl);
    }
  };

  const Error = errorState && (
    <div
      className={`${styles.error} ${styles.response}`}
      data-testid="urlShortener-error"
    >
      There is something wrong with the system
    </div>
  );

  const ResultNote = shortenUrl && (
    <div className={`${styles.result} ${styles.response}`}>
      <span>shorten url: </span>
      <div
        className={styles.shortenUrl}
        data-testid="urlShortener-result"
      >
        {shortenUrl}
      </div>
    </div>
  );

  return (
    <div className={styles.container}>
      <h1>url shortener</h1>
      <div className={styles.inputPanel}>
        <input
          type="text"
          className={styles.input}
          onChange={onInputChange}
          placeholder="original URL"
        />
        <input
          type="submit"
          value="Submit"
          className={styles.button}
          onClick={submitHandler}
          data-testid="urlShortener-button"
        />
      </div>
      {Error}
      {ResultNote}
    </div>
  );
};

export default withStyles(styles)(UrlShortener);
