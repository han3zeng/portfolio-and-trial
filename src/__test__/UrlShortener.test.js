import React from 'react';
import { create } from 'react-test-renderer';
import { render, fireEvent, waitForElement } from '@testing-library/react'
import UrlShortener from '../components/UrlShortener';
import { createShortUrl } from '../utils/api';

jest.mock('../utils/api');
createShortUrl
  .mockImplementationOnce(async () => {
    const res = await Promise.resolve({
      data: {
        shortenUrl: 'test-shorten-url',
      },
    });
    return res;
  })
  .mockImplementationOnce(async () => {
    const res = await Promise.resolve({
      error: new Error('this is error testing obj'),
    });
    return res;
  });

describe('<UrlShortener />', () => {
  it('snapshot check', () => {
    const root = create(<UrlShortener />);
    expect(root.toJSON()).toMatchSnapshot();
  });

  it('should be successful scenario result after click', async () => {
    const { getByTestId } = render(<UrlShortener />);
    const button = getByTestId('urlShortener-button');
    fireEvent.click(button);
    const resultDiv = await waitForElement(() => getByTestId('urlShortener-result'));
    expect(resultDiv.textContent).toBe('test-shorten-url');
  });

  it('should be fail scenario result after click', async () => {
    const { getByTestId } = render(<UrlShortener />);
    const button = getByTestId('urlShortener-button');
    fireEvent.click(button);
    const resultDiv = await waitForElement(() => getByTestId('urlShortener-error'));
    expect(resultDiv.textContent).toBe('There is something wrong with the system');
  });
});

/*
  react-testing-library: support react hook
  https://testing-library.com/docs/react-testing-library/example-intro
  https://github.com/testing-library/react-testing-library/issues/379
*/
