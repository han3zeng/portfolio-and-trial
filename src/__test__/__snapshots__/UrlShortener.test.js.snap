// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UrlShortener /> snapshot check 1`] = `
<div
  className="container"
>
  <h1>
    url shortener
  </h1>
  <div
    className="inputPanel"
  >
    <input
      className="input"
      onChange={[Function]}
      placeholder="original URL"
      type="text"
    />
    <input
      className="button"
      data-testid="urlShortener-button"
      onClick={[Function]}
      type="submit"
      value="Submit"
    />
  </div>
</div>
`;
